docs/features/source-generators.cookbook.md

```diff
@@ -0,0 +1,250 @@
+            {
+                 // find the class to augment
+                var classToAugment = syntaxTree.GetRoot().DescendantNodes()
+                 .OfType<ClassDeclarationSyntax>()
+                 .Where(c => c.Name.ToString() == "UserClass")
+                 .Single();
+
+                var sourceText = SourceText.From($@"
+public partial class {classToAugment.Identifier.ToString()}
+{
+   private void GeneratedMethod()
+   {
+       // generated code
+   }
+}");
+                context.AddSource("myGeneratedFile.cs", sourceText);
+            }
+        }
+    }
+
+```
+
+### Participate in the IDE experience
+
+**User scenario:** As a generator author I want to be able to interactively regenerate code as the user is editing files.
+
+**Solution:** We expect there to be an an opt-in set of interactive interfaces that can be implemented to allow for progressively more complex generation strategies.
+It is anticipated there will be a mechanism for providing symbol mapping for lighting up features such a 'Find all references'.
+
+```csharp
+public class InteractiveGenerator : ISourceGenerator, IAdditionalFilesChangedGenerator
+    {
+        public void Execute(SourceGeneratorContext context)
+        {
+            // generators must always support a total generation pass
+        }
+
+        public void OnAdditionalFilesChanged(AdditionalFilesChangedContext context)
+        {
+            // determine which file changed, and if it affects this generator
+            // regenerate only the parts that are affected by this change.
+        }
+    }
+```
+
+### INotifyPropertyChanged
+
+**User scenario:** As a generator author I want to be able to implement the `INotifyPropertyChanged` pattern automatically for a user.
+
+**Solution:** The design tenant 'Explicitly additive only' seems to be at direct odds with the ability to implement this, and appears to call for user code modification.
+However we can instead take advantage of explict fields and instead of *editing* the users properties, directly provide them for listed fields.
+
+**Example:**
+
+Given a user class such as:
+
+```csharp
+public partial class UserClass
+{
+    [AutoNotify]
+    private bool _boolProp;
+
+    [AutoNotify(propertyName: "Count")]
+    private int _intProp;
+}
+```
+
+A generator could produce the following:
+
+```csharp
+using System.ComponentModel;
+
+public partial class UserClass : INotifyPropertyChanged
+{
+    public bool BoolProp
+    {
+        get => _boolProp;
+        set
+        {
+            _boolProp = value;
+            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs("UserBool"));
+        }
+    }
+
+    public int Count
+    {
+        get => _intProp;
+        set
+        {
+            _intProp = value;
+            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs("Count"));
+        }
+    }
+
+    public event PropertyChangedEventHandler PropertyChanged;
+}
+
+```
+
+### Serialization
```

Comment:
Definitely this one needs to get fleshed out. It's probably the most critical one to have a good definition for.